packages/assets-controllers/src/MultichainTokensRatesController/MultichainTokensRatesController.ts

```diff
@@ -0,0 +1,391 @@
+   */
+  #getAccount(accountId: string): InternalAccount | undefined {
+    return this.#listAccounts().find((account) => account.id === accountId);
+  }
+
+  /**
+   * Handles the removal of an account by deleting its conversion rates.
+   *
+   * @param accountId - The ID of the removed account.
+   */
+  async #handleOnAccountRemoved(accountId: string): Promise<void> {
+    if (accountId in this.state.conversionRates) {
+      this.update((state: Draft<MultichainTokensRatesControllerState>) => {
+        delete state.conversionRates[accountId];
+      });
+    }
+  }
+
+  /**
+   * Updates the token conversion rates for a given account.
+   *
+   * This method acquires a mutex lock to ensure thread safety.
+   *
+   * @param accountId - The account ID for which to update rates.
+   * @returns A promise that resolves when the rates have been updated.
+   */
+  async updateTokensRates(accountId: string): Promise<void> {
+    const releaseLock = await this.#mutex.acquire();
+
+    return (async () => {
+      if (!this.isActive) {
+        return;
+      }
+
+      const account = this.#getAccount(accountId);
+
+      if (!account?.metadata.snap) {
+        return;
+      }
+
+      // TODO: Remove this condition when additional Snap support is implemented.
+      if (account.type !== 'solana:data-account') {
+        return;
+      }
+
+      // Retrieve assets from the assets controller.
+      const assets = await this.#getAssetsList(
+        accountId,
+        account.metadata.snap.id as SnapId,
+      );
+
+      const conversions = assets.map((asset) => ({
+        from: asset,
+        to: SWIFT_ISO4217_USD,
+      }));
+
+      const accountRates = await this.#handleSnapRequest({
+        snapId: account.metadata.snap.id as SnapId,
+        handler: 'onAssetsConversion' as HandlerType,
+        conversions,
+      });
+
+      // Update the state with new conversion rates.
+      this.update((state: Draft<MultichainTokensRatesControllerState>) => {
+        state.conversionRates[accountId] = accountRates.conversionRates;
+      });
+    })().finally(() => {
+      releaseLock();
+    });
+  }
+
+  /**
+   * Forwards a Snap request to the SnapController.
+   *
+   * @param args - The request parameters.
+   * @param args.snapId - The ID of the Snap.
+   * @param args.handler - The handler type.
+   * @param args.conversions - The asset conversions.
+   * @returns A promise that resolves with the account rates.
+   */
+  async #handleSnapRequest({
+    snapId,
+    handler,
+    conversions,
+  }: {
+    snapId: SnapId;
+    handler: HandlerType;
+    conversions: { from: CaipAssetTypeOrId; to: string }[];
+  }): Promise<{ conversionRates: AccountConversionRates }> {
+    return this.messagingSystem.call('SnapController:handleRequest', {
+      snapId: snapId as SnapId,
+      origin: 'metamask',
+      handler,
+      request: {
+        id: '4dbf133d-9ce3-4d3f-96ac-bfc88d351046',
```

Comment:
Id shouldn't be hardcoded, I believe this should be a random uuid?

Reply:
yes same here , i'll use uuid library to generate random one

Reply:
You shouldn't need to set a request ID as it's set by the `SnapController`: https://github.com/MetaMask/snaps/blob/1966bf456c8e9038da4c688882ade497e329d5d6/packages/snaps-controllers/src/snaps/SnapController.ts#L3377